js/copy-service/Evaluator/Evaluator.ts

```diff
@@ -0,0 +1,109 @@
+import _ from 'lodash';
+
+import CopyService, { AST } from '../CopyService';
+import ErrorHandler from '../ErrorHandler/ErrorHandler';
+import IntlCopyService from '../IntlCopyService';
+import Substitutions from '../Substitutions/Substitutions';
+
+/**
+ * Provides an interface to recursively generate copy evaluated with substitutions.
+ *
+ * @interface
+ */
+class Evaluator {
+  /**
+   * The copy service used for evaluation.
+   *
+   * @type {CopyService|IntlCopyService}
+   */
+  copyService: CopyService | IntlCopyService;
+
+  /**
+   * A cache for previously evaluated cope.
+   *
+   * @type {WeakMap}
+   */
+  evaluationCache: WeakMap<object, unknown>;
```

Comment:
We really want to avoid `unknown` unless absolutely necessary. In subclasses, we cache the evaluated copy, which will be of type `string` or `React.Element` depending on which evaluator is chosen. However, consuming repos can define their own Evaluator subclasses evaluating copy to pretty much anything, so we should define the value as `any` and let the subclasses override the type to a more specific type.

```suggestion
  evaluationCache: WeakMap<SyntaxNode, any>;
```